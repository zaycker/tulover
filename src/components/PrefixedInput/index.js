import React, { useState, useCallback, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.module.css';

const useFullyControlledInput = (value, onChange, pattern) => {
    const [stateValue, setStateValue] = useState(value);

    useEffect(() => {
        if (stateValue !== value) {
            setStateValue(value);
        }
    }, [value]); // eslint-disable-line react-hooks/exhaustive-deps

    const onChangeCallback = useCallback((e) => {
        const patternRegExp = new RegExp(`(${pattern})`);
        const match = e.target.value.match(patternRegExp);
        setStateValue(match ? match[0] : '');
        onChange && onChange(e);
    }, [onChange, pattern]);

    return [stateValue, onChangeCallback];
};

export default function PrefixedInput({ value, onChange, prefix, autoFocus, pattern, ...props }) {
    const [stateValue, onInputChange] = useFullyControlledInput(value, onChange, pattern);
    const inputRef = useRef(null);

    // autoFocus won't work here cause input rendered once and becomes visible with parent slide many times
    // that's why refs and focus() by hand
    useEffect(() => {
        if (autoFocus && inputRef.current) {
            inputRef.current.focus();
        }
    }, [autoFocus]);

    return (
        <label className={styles['prefixed-input__label']}>
            {!!stateValue && (
                <span className={styles['prefixed-input__sign']}>{prefix}{stateValue}</span>
            )}
            <input
                type="text"
                className={styles['prefixed-input__input']}
                value={stateValue}
                onChange={onInputChange}
                ref={inputRef}
                {...props}
            />
        </label>
    );
}

PrefixedInput.propTypes = {
    onChange: PropTypes.func,
    prefix: PropTypes.string,
    value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
    ]),
    pattern: PropTypes.string,
};

PrefixedInput.defaultProps = {
    onChange: Function.prototype,
    prefix: '',
    value: '',
    pattern: '.*',
};
