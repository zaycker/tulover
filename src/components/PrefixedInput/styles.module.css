.prefixed-input__label {
  position: relative;
}

.prefixed-input__sign,
.prefixed-input__input {
  font-size: 2em;
  line-height: 1.2;
  padding: 0;
  margin: 0;
  border: 0;
  outline: 0;
  background: none;
  color: #fff;
  max-width: 200px;
}

.prefixed-input__sign {
  position: absolute;
  right: 0;
}

.prefixed-input__input {
  position: relative;
  z-index: 1;
  text-align: right;
}

.prefixed-input__input::placeholder {
  color: rgba(255, 255, 255, 0.2);
  font-weight: 300;
}