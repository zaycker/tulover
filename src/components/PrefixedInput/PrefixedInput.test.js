import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import PrefixedInput from './index';

it('render correctly PrefixedInput component', () => {
    const PrefixedInputComponent = renderer.create(
        <PrefixedInput />
    ).toJSON();
    expect(PrefixedInputComponent).toMatchSnapshot();
});

it('correctly validates values by pattern on change', () => {
    const wrapper = mount(
        <PrefixedInput pattern="^\d+\.?\d{0,2}$" />
    );

    wrapper.find('.prefixed-input__input').simulate('change', { target: { value: '123.45' }});
    expect(wrapper.find('.prefixed-input__input').props().value).toBe('123.45');

    wrapper.find('.prefixed-input__input').simulate('change', { target: { value: '12abc' }});
    expect(wrapper.find('.prefixed-input__input').props().value).toBe('');
});