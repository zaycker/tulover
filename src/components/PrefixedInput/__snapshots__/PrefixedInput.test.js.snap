// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly PrefixedInput component 1`] = `
<label
  className="prefixed-input__label"
>
  <input
    className="prefixed-input__input"
    onChange={[Function]}
    type="text"
    value=""
  />
</label>
`;
